### PR TITLE
 (PDB-2791) Add tests and new open/store exception

### PR DESCRIPTION
**Pull request by rbrw**

### CODE REVIEW

**Comment by rbrw** [comment]
RFC: regarding the new "orphan" exception

**Comment by rbrw** [comment]
This is based on #11 and shouldn't be merged yet.

File: src/puppetlabs/stockpile/queue.clj [hunk 1]
```diff
@@ -80,26 +80,36 @@
...
   [src dest sync-parent?]
   (atomic-move src dest)
   (when sync-parent?
     (fsync (.getParent (as-path dest)) true)))
 
+(defn- delete-if-exists [path]
+  ;; Solely exists for error handling tests
+  (Files/deleteIfExists path))
+
+(defn- write-stream [stream dest]
```
Comment (senior) [inline src/puppetlabs/stockpile/queue.clj]: This needs type hints for stream and dest

File: src/puppetlabs/stockpile/queue.clj [hunk 2]
```diff
@@ -108,12 +118,15 @@
...
-        ;; This approach will be revisited/revised after we discuss
-        ;; the alternatives a bit further.
         (try
-          (Files/deleteIfExists tmp)
-          (catch Exception ex
-            true))
+          (delete-if-exists tmp)
+          (catch Exception del-ex
+            (throw
+             (ex-info (str "unable to delete temp file " tmp " after error")
```
Comment (senior) [inline src/puppetlabs/stockpile/queue.clj:125]: We talked about this Monday and I know you're working up some changes here, I'll hold off on the rest of this review until that's done.

**Review by senior — CHANGES_REQUESTED** [review]
You've already got some changes coming, I'll look at it again once the related PR goes in.

File: src/puppetlabs/stockpile/queue.clj [hunk 3]
```diff
@@ -265,7 +279,9 @@
   subtracting the 20 (UTF-8 encoded Basic Latin block) bytes reserved
   for internal use, there may be up to 235 bytes available for the
   metadata.  Of course how many Unicode characters that will allow
-  depends on their size when converted to UTF-8."
+  depends on their size when converted to UTF-8.  On error, may throw
+  the ::path-cleanup-failure-after-error exception documented in
+  durably-establish."
```
Comment (senior) [inline src/puppetlabs/stockpile/queue.clj]: This public function is referencing the docs in a private function, that seems wrong to me as usually private functions don't have their docs published. Is there a list of all of the different exceptions that need to be caught by callers of this function?

Comment (rbrw) [reply src/puppetlabs/stockpile/queue.clj]: Good point; I've included the docs in the create and store docstrings and elaborated on how we think callers might want to handle invocations.

**Pull request merged**